### DISCUSSION TITLE
TypeError: store.getState is not a function

### DISCUSSION THREAD

**Discussion by ikotsov** [pmndrs/zustand#1709] [Q&A] [answered]
Hello,

I am following [this guide](https://docs.pmnd.rs/zustand/guides/testing) in order to achieve resetting state between my tests.

My project is in React Native and I am using jest for testing.

Unfortunately, it is not clear in the docs how should I use the store created in `zustand.ts` file in `__mocks__` folder. I suppose that just by adding it, it will work automatically. Is this correct?

The code of `zustand.ts` file is this:
```
// eslint-disable-next-line import/no-extraneous-dependencies
import { act } from '@testing-library/react-native';
import { StateCreator } from 'zustand';

const { create: actualCreate } = jest.requireActual('zustand');

// a variable to hold reset functions for all stores declared in the app
const storeResetFns = new Set<() => void>();

// when creating a store, we get its initial state, create a reset function and add it in the set
export const create = <S>(createState: StateCreator<S>) => {
  const store = actualCreate(createState);
  const initialState = store.getState();
  storeResetFns.add(() => store.setState(initialState, true));
  return store;
};

// Reset all stores after each test run
beforeEach(async () => {
  await act(() =>
    storeResetFns.forEach(resetFn => {
      resetFn();
    })
  );
});

```

I get this error: TypeError: store.getState is not a function

<img width="573" alt="Screenshot 2023-03-23 at 12 04 14 PM" src="https://user-images.githubusercontent.com/15989223/227170160-92f7e363-c0bd-49e2-b35d-a2805abedcc3.png">

My store and tests setup is like this:

```
type State = {
  passengers: number;
  actions: {
    addPassenger: () => void;
    removePassenger: () => void;
  };
};

export const useStore = create<State>()(set => ({
  passengers: 1,
  actions: {
    addPassenger: () => set(state => ({ passengers: state.passengers + 1 })),
    removePassenger: () => set(state => ({ passengers: state.passengers - 1 }))
  }
}));

export const usePassengers = () => useStore(state => state.passengers);

// ... and then in my test file

test('1 passenger has been added as a default selection', () => {
  const { result } = renderHook(() => usePassengers());

  expect(result.current).toBe(1);
});


```

I also looked at [previous issues](https://github.com/pmndrs/zustand/issues?q=is%3Aissue+TypeError%3A+store.getState+is+not+a+function+is%3Aclosed), but there is not any obvious solution to this.

Based on [this proposal](https://github.com/pmndrs/zustand/issues/271), alternatively I have to do a manual reset for each one of the test files I write, which is not ideal.

Is there any idea how I can fix this type error?

**Comment by dai-shi** [comment, marked as answer]
You want to see what's `store` with console.log?

```js
const store = actualCreate(createState);
console.log(store);
```

**Reply by ikotsov** [reply]
> You want to see what's `store` with console.log?
> 
> ```js
> const store = actualCreate(createState);
> console.log(store);
> ```

Yes of course, it logs the below:
```
  console.log
    [Function: createImpl]
```

**Reply by dai-shi** [reply]
Okay, your mock and the usage doesn't match. I wonder if we have discussions (and possible doc fixes) for TypeScript testing usage.

For now, this should fix.
```diff
- export const useStore = create<State>()(set => ({
+ export const useStore = create<State>(set => ({
```

**Reply by dai-shi** [reply]
Well, the better is to fix the mock.

```ts
export const create = <S>() => (createState: StateCreator<S>) => {
  const store = actualCreate(createState);
  const initialState = store.getState();
  storeResetFns.add(() => store.setState(initialState, true));
  return store;
};
```

@dbritto-dev @sewera @chrisk-7777 Didn't we have discussions for this?

**Reply by dbritto-dev** [reply]
@dai-shi yes, we did haha but we ended up just adding a note

**Reply by ikotsov** [reply]
Good catch @dai-shi! This was the reason why it failed. After doing the modification you proposed, It's working.

It's not directly connected with this discussion, but I am including it for historical reasons. I also encountered a warning with the "await act" I was doing at the end of the `zustand.ts` file and before each test.

This was the warning I received:

<img width="685" alt="Screenshot 2023-03-24 at 8 43 57 AM" src="https://user-images.githubusercontent.com/15989223/227446222-9932f72f-157b-4c1a-917a-f985a220abaa.png">

I removed the await in order for it to run without warnings.

Thank you a lot for the immediate answer. 🙏

I think it would be valuable to update [this](https://docs.pmnd.rs/zustand/guides/testing) file so that it is clearer how to handle the typescript case.

**Reply by tonycassara** [reply]
Hi I'm having the same issue in a codebase where we're mixing regular zustand store and store that's been created using URL Hash. The issue I think is that the mocked shape won't match both cases. I removed the curried `()` used when we `create` with a type:

```
export const useUrlHashFilterParams = create<FilterParams>(
  persist(
    (set) => ({ ... }),
    {
      name: "url-hash-filters",
      storage: createJSONStorage(() => hashStorage),
    },
  ),
);
```

And the store continues to work but I get errors:

```
Argument of type 'StateCreator<FilterParams, [], [["zustand/persist", unknown]]>' is not assignable to parameter of type 'StateCreator<FilterParams, [], []>'.
  Type 'StateCreator<FilterParams, [], [["zustand/persist", unknown]]>' is not assignable to type '{ $$storeMutators?: [] | undefined; }'.
    Types of property '$$storeMutators' are incompatible.
      Type '[["zustand/persist", unknown]] | undefined' is not assignable to type '[] | undefined'.
        Type '[["zustand/persist", unknown]]' is not assignable to type '[]'.
          Source has 1 element(s) but target allows only 0.ts(2345)
```

Also in the mocked `zustand` I get the same error:

```
Argument of type '[]' is not assignable to parameter of type 'string | { type: unknown; } | undefined'.
```

But we do a little something different in our mock:

```
export const create = <
  T extends unknown,
  Mos extends [StoreMutatorIdentifier, unknown][] = [],
>(
  createState: StateCreator<T, [], Mos>,
) => {
  const store = actualCreate(createState);
  const initialState = store.getState();
  storeResetFns.add(() => store.setState(initialState, true));
  return store;
};
```
Any ideas how I should go about this? Do I need to add an overload to the mocked `zustand` store?

**Reply by dai-shi** [reply]
@dbritto-dev should be able to help?

**Reply by dbritto-dev** [reply]
@dai-shi sure. @tonycassara would you mind creating an new discussion thread and give me a codesandbox demo to replicate the issue?

**Reply by tonycassara** [reply]
@dai-shi will do tomorrow thank you
